File: problem/Largest_Magic_Square/largestMagicSquare.c
#include <stdlib.h>
#include <stdbool.h>

int largestMagicSquare(int **grid, int gridSize, int *gridColSize)
{
    int r = gridSize;
    int c = gridColSize[0];

    int **rowSum = (int **)malloc(r * sizeof(int *));
    for (int i = 0; i < r; i++)
    {
        rowSum[i] = (int *)calloc(c + 1, sizeof(int));
        for (int j = 0; j < c; j++)
        {
            rowSum[i][j + 1] = rowSum[i][j] + grid[i][j];
        }
    }

    int **colSum = (int **)malloc(c * sizeof(int *));
    for (int j = 0; j < c; j++)
    {
        colSum[j] = (int *)calloc(r + 1, sizeof(int));
        for (int i = 0; i < r; i++)
        {
            colSum[j][i + 1] = colSum[j][i] + grid[i][j];
        }
    }

    int **diag = (int **)malloc((r + 1) * sizeof(int *));
    int **antidiag = (int **)malloc((r + 1) * sizeof(int *));
    for (int i = 0; i <= r; i++)
    {
        diag[i] = (int *)calloc(c + 1, sizeof(int));
        antidiag[i] = (int *)calloc(c + 1, sizeof(int));
    }

    for (int i = 0; i < r; i++)
    {
        for (int j = 0; j < c; j++)
        {
            diag[i + 1][j + 1] = diag[i][j] + grid[i][j];
            antidiag[i + 1][j] = antidiag[i][j + 1] + grid[i][j];
        }
    }

    int min_dim = r < c ? r : c;
    for (int k = min_dim; k > 1; k--)
    {
        for (int i = 0; i <= r - k; i++)
        {
            for (int j = 0; j <= c - k; j++)
            {
                int sum = diag[i + k][j + k] - diag[i][j];
                int antiSum = antidiag[i + k][j] - antidiag[i][j + k];

                if (sum != antiSum)
                    continue;

                bool match = true;
                for (int m = 0; m < k; m++)
                {
                    int rSum = rowSum[i + m][j + k] - rowSum[i + m][j];
                    int cSum = colSum[j + m][i + k] - colSum[j + m][i];
                    if (rSum != sum || cSum != sum)
                    {
                        match = false;
                        break;
                    }
                }

                if (match)
                    return k;
            }
        }
    }

    return 1;
}

File: problem/Largest_Magic_Square/largestMagicSquare.py
class Solution(object):
    def largestMagicSquare(self, grid):
        r, c = len(grid), len(grid[0])

        rowSum = [[0] * (c + 1) for _ in range(r)]
        for i in range(r):
            for j in range(c):
                rowSum[i][j+1] = rowSum[i][j] + grid[i][j]

        colSum = [[0] * (r + 1) for _ in range(c)]
        for j in range(c):
            for i in range(r):
                colSum[j][i+1] = colSum[j][i] + grid[i][j]

        diag = [[0] * (c + 1) for _ in range(r + 1)]
        antidiag = [[0] * (c + 1) for _ in range(r + 1)]

        for i in range(r):
            for j in range(c):
                diag[i+1][j+1] = diag[i][j] + grid[i][j]
                antidiag[i+1][j] = antidiag[i][j+1] + grid[i][j]

        def isMagic(k):
            for i in range(r - k + 1):
                for j in range(c - k + 1):
                    Sum = diag[i+k][j+k] - diag[i][j]
                    antiD = antidiag[i+k][j] - antidiag[i][j+k]
                    
                    if Sum != antiD:
                        continue
                    
                    match = True
                    for m in range(k):
                        row_val = rowSum[i+m][j+k] - rowSum[i+m][j]
                        col_val = colSum[j+m][i+k] - colSum[j+m][i]
                        
                        if row_val != Sum or col_val != Sum:
                            match = False
                            break
                    
                    if match:
                        return True
            return False

        for k in range(min(r, c), 1, -1):
            if isMagic(k): 
                return k
        
        return 1